libbeat/publisher/queue/diskqueue/writer_loop.go

```diff
@@ -0,0 +1,239 @@
+	return &writerLoop{
+		logger:   logger,
+		settings: settings,
+
+		requestChan:  make(chan writerLoopRequest, 1),
+		responseChan: make(chan writerLoopResponse),
+	}
+}
+
+func (wl *writerLoop) run() {
+	for {
+		block, ok := <-wl.requestChan
+		if !ok {
+			// The request channel is closed, we are done
+			return
+		}
+		bytesWritten := wl.processRequest(block)
+		wl.responseChan <- writerLoopResponse{bytesWritten: bytesWritten}
+	}
+}
+
+// processRequest writes the frames in the given request to disk and returns
+// the number of bytes written to each segment, in the order they were
+// encountered.
+func (wl *writerLoop) processRequest(request writerLoopRequest) []int64 {
+	// retryWriter wraps the file handle with timed retries.
+	// retryWriter.Write is guaranteed to return only if the write
+	// completely succeeded or the queue is being closed.
+	retryWriter := callbackRetryWriter{retry: wl.retryCallback}
+
+	// We keep track of how many frames are written during this request,
+	// and send the associated ACKs to the queue / producer listeners
+	// in a batch at the end (since each ACK call can involve a round-trip
+	// to the registry).
+	totalACKCount := 0
+	producerACKCounts := make(map[*diskQueueProducer]int)
+
+	var bytesWritten []int64    // Bytes written to all segments.
+	curBytesWritten := int64(0) // Bytes written to the current segment.
+outerLoop:
+	for _, frameRequest := range request.frames {
+		// If the new segment doesn't match the last one, we need to open a new
+		// file handle and possibly clean up the old one.
+		if wl.currentSegment != frameRequest.segment {
+			wl.logger.Debugf(
+				"Creating new segment file with id %v\n", frameRequest.segment.id)
+			if wl.outputFile != nil {
+				// Try to sync to disk, then close the file.
+				wl.outputFile.Sync()
+				wl.outputFile.Close()
+				wl.outputFile = nil
+				// We are done with this segment, add the byte count to the list and
+				// reset the current counter.
+				bytesWritten = append(bytesWritten, curBytesWritten)
+				curBytesWritten = 0
+			}
+			wl.currentSegment = frameRequest.segment
+			file, err := wl.currentSegment.getWriterWithRetry(
+				wl.settings, wl.retryCallback)
+			if err != nil {
+				// This can only happen if the queue is being closed; abort.
+				break
+			}
+			wl.outputFile = file
+		}
+		// Make sure our writer points to the current file handle.
+		retryWriter.wrapped = wl.outputFile
+
+		// We have the data and a file to write it to. We are now committed
+		// to writing this block unless the queue is closed in the meantime.
+		frameSize := uint32(frameRequest.frame.sizeOnDisk())
+
+		// The Write calls below all pass through retryWriter, so they can
+		// only return an error if the write should be aborted. Thus, all we
+		// need to do when we see an error is break out of the request loop.
+		err := binary.Write(retryWriter, binary.LittleEndian, frameSize)
+		if err != nil {
```

Comment:
Some debug logs would be nice when we get an error after any of these `Write` functions.

Reply:
They are logged, see `retryCallback` below (all writes here are wrapped by the callback writer, which routes errors through that helper).